Read Parquet rowgroups and columns in parallel

jpdna opened the pull request:
These changes parallelize the reading of Parquet files by both rowgroup and column.

In this PR's implementation, the number of remote/parallel jobs is exactly defined by the on-disk structure of the parquet file: num_rowgroups * num_columns.
Each modin block is one rowgroup, column pair.
Most likely this should be optimized based on available compute resources.

Benchmarks:
In a test loading a 4.4GB parquet file with 5 columns, on a 32 CPU machine, this implementation ran for 35 seconds while the current Master version runs for 165 seconds.
Watching Top is was clear that the previous version was limited to using 5 CPU, one for each column, while this new version was making use of all available CPUs.

Todo:
- Discuss adding more adaptive parallelism  based on user compute resources,.
- more benchmarks
- add tests / needs further validation that the schema created works with all modin/panda functions

devin-petersohn commented on modin/engines/ray/generic/io.py:
```diff
@@ -103,6 +103,8 @@ class RayIO(BaseIO):
     #
     # Signature: (num_splits, sql, con, index_col, kwargs)
 
+    from pyarrow.parquet import ParquetFile
```
This import should be moved inside the relevant functions/methods that require it.

devin-petersohn commented on modin/engines/ray/generic/io.py:
```diff
@@ -134,6 +136,7 @@ def read_parquet(cls, path, engine, columns, **kwargs):
             ]
         num_partitions = cls.frame_mgr_cls._compute_num_partitions()
         num_splits = min(len(columns), num_partitions)
+    
```
Delete extra newline

devin-petersohn commented on modin/engines/ray/generic/io.py:
```diff
@@ -144,24 +147,48 @@ def read_parquet(cls, path, engine, columns, **kwargs):
-        ).T
+     
+        pf = ParquetFile(path)
+        column_names = [
```
We should already have `columns` here, do we need to compute this separately?

devin-petersohn commented on modin/engines/ray/generic/io.py:
```diff
@@ -144,24 +147,48 @@ def read_parquet(cls, path, engine, columns, **kwargs):
+              ]
+
+        num_row_groups = pf.metadata.num_row_groups 
+     
```
Try to reduce vertical whitespace throughout the PR.

devin-petersohn commented on modin/engines/ray/generic/io.py:
```diff
@@ -144,24 +147,48 @@ def read_parquet(cls, path, engine, columns, **kwargs):
                 for row in blk_partitions[:-1]
             ]
         )
+
```
Also remove extra newline.

devin-petersohn commented on modin/engines/ray/generic/io.py:
```diff
@@ -144,24 +147,48 @@ def read_parquet(cls, path, engine, columns, **kwargs):
+        num_row_groups = pf.metadata.num_row_groups 
+     
+
+        blocks_step1 =  [ 
```
We may need to rethink the idea that each `row_group` + `column` combination returns the `index` or the length of the `index`. This is creating multiple copies.

It would be good to try to get this to work with 2D list comprehension. Example:

```python
blocks = [
    [
        cls.read_parquet_remote_task._remote(
            args=(path, col_partitions[i], num_splits=1, row_group=j, kwargs),
            num_return_vals=1 if j != num_row_groups - 1 else 2
        ) for i in range(len(column_splits))
    ] 
    for j in range(num_row_groups)
]
```

Of course we would have to add the logic for handling the last column group, but we should take advantage of the `col_partitions` variable created above to avoid over-splitting the data.

jpdna replied:
When you say above, "this creates multiple copies",  can you describe for me in more details what object/data exactly you are referring to that is getting copied multiple times?  Do you mean the redundant length numbers ( which are indeed the same for all columns), though in the case here where it is just and integer, not actually and index, seems not too expensive.

devin-petersohn replied:
Right, the index (which is an integer in the common case). This is expensive because there are fixed overheads to each `ray.put`. We want to avoid putting data in the object store that is redundant/never needed.

devin-petersohn replied:
Let's try that. We can always iterate.

devin-petersohn commented on modin/engines/ray/pandas_on_ray/io.py:
```diff
@@ -49,9 +52,15 @@ def _read_parquet_columns(path, columns, num_splits, kwargs):  # pragma: no cove
+    if(rowgroup>=0):
+      pf = ParquetFile(path)
+      df = pf.read_row_group(rowgroup, columns, **kwargs).to_pandas()
+      return _split_result_for_readers(0, 1, df) + [len(df.index)] 
```
If `num_splits=1`, this will just return the `df` itself. It may not work unless you wrap it in a list yourself. We don't need to split in this case so we can just return it like this:

```python
return [df] + [len(df.index)]
```

jpdna commented:
I didn't find any particularly good way to avoid returning both the df and the row_number from each remote call processing a block while still also retaining the elegance of using a 2d list comprehension.  I did however remove redundant `ray.put` of the final row count, by just appending `blocks` with a list with repeated references to a single object_id containing the summed row count.

devin-petersohn commented:
Ok, that is fine. There are still some unaddressed comments in your recent push and the tests are failing for `io`. Once those are addressed I will take another pass.

jpdna commented:
Ah- I'd rebased but not committed my latest changes, hence the unaddressed comments.
I'm trying anyway to run the tests locally before I push again, but I get an error both for this PR and on Master  at d726c5e7c5662  when I run `pytest modin/pandas/test`
The output from running test on Master branch  is below:

Can you point me to the correct command to run tests locally?
```
(modindev7) [paschalj@mynode modin]$ pytest modin/pandas/test
========================================================== test session starts ==========================================================
platform linux -- Python 3.6.8, pytest-4.4.1, py-1.8.0, pluggy-0.9.0
rootdir: /myfullpath/users/paschalj/modin/april17_master/modin
collecting 97 items / 2 errors / 95 selected                                                                                            INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pytest-4.4.1-py3.6.egg/_pytest/main.py", line 209, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pytest-4.4.1-py3.6.egg/_pytest/main.py", line 248, in _main
INTERNALERROR>     config.hook.pytest_collection(session=session)
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pluggy-0.9.0-py3.6.egg/pluggy/hooks.py", line 289, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pluggy-0.9.0-py3.6.egg/pluggy/manager.py", line 68, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pluggy-0.9.0-py3.6.egg/pluggy/manager.py", line 62, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pluggy-0.9.0-py3.6.egg/pluggy/callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pluggy-0.9.0-py3.6.egg/pluggy/callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pluggy-0.9.0-py3.6.egg/pluggy/callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pytest-4.4.1-py3.6.egg/_pytest/main.py", line 258, in pytest_collection
INTERNALERROR>     return session.perform_collect()
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pytest-4.4.1-py3.6.egg/_pytest/main.py", line 495, in perform_collect
INTERNALERROR>     self.config.pluginmanager.check_pending()
INTERNALERROR>   File "/mydata/paschalj//conda/envs/modindev7/lib/python3.6/site-packages/pluggy-0.9.0-py3.6.egg/pluggy/manager.py", line 251, in check_pending
INTERNALERROR>     % (name, hookimpl.plugin),
INTERNALERROR> pluggy.manager.PluginValidationError: unknown hook 'pytest_benchmark_update_commit_info' in plugin <module 'conftest' from '/myfullpath/users/paschalj/modin/april17_master/modin/modin/pandas/test/performance-tests/conftest.py'>

================================================= 2 wa

```

devin-petersohn commented:
The performance tests should be ignored when running pytest. The relevant failures are in `test_io.py`.

jpdna commented:
Can you tell me the command to run that runs `test_io.,py` locally?
On master I tried: `pytest modin/pandas/test/test_io.py` but still gets lots of errors where I am sure Master should pass.

devin-petersohn commented:
You may be missing dependencies for all of the I/O tests.

`pip install -r requirements.txt`

Try that first.

jpdna commented:
yup, thanks  - test passed on master now, I'll now check my PR before pushing again.

jpdna commented:
Tests are still running on Travis, locally I get some errors related to picking like pasted below.
However, testing manually using ipython, the loaded parquet files produce a dataset with same number of rows, and when exported to csv have the same content so far, in both original pandas and Modin with this PR.  If there is something serious that my PR breaks wrt to pickling, I'm very surprised that it seems to work on my manual test dataset, please let me know if you have any thoughts on the test errors below.

An example of error when running `pytest modin/pandas/test/test_io.py` that is unique to this PR is:
```
pytest modin/pandas/test/test_io.py
====================================================================================== test session starts =======================================================================================
platform linux -- Python 3.6.8, pytest-3.9.3, py-1.8.0, pluggy-0.9.0
rootdir: /users/paschalj/modin/v10/modin, inifile:
plugins: xdist-1.27.0, forked-1.0.2, cov-2.6.1
collected 50 items                                                                                                                                                                               

modin/pandas/test/test_io.py FF..s.........s...F.........s.s.X........s........                                                                                                            [100%]

============================================================================================ FAILURES ============================================================================================
_______________________________________________________________________________________ test_from_parquet ________________________________________________________________________________________

self = <ray.cloudpickle.cloudpickle.CloudPickler object at 0x2aaab67175c0>, obj = <function PandasQueryCompiler._prepare_method.<locals>.helper at 0x2aaab671c7b8>

    def dump(self, obj):
        self.inject_addons()
        try:
>           return Pickler.dump(self, obj)

/data/paschalj/conda/envs/modindev10/lib/python3.6/site-packages/ray-0.6.2-py3.6-linux-x86_64.egg/ray/cloudpickle/cloudpickle.py:268: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ray.cloudpickle.cloudpickle.CloudPickler object at 0x2aaab67175c0>, obj = <function PandasQueryCompiler._prepare_method.<locals>.helper at 0x2aaab671c7b8>

    def dump(self, obj):
        """Write a pickled representation of obj to the open file."""
        # Check whether Pickler was initialized correctly. This is
        # only needed to mimic the behavior of _pickle.Pickler.dump().
        if not hasattr(self, "_file_write"):
            raise PicklingError("Pickler.__init__() was not called by "
                                "%s.__init__()" % (self.__class__.__name__,))
        if self.proto >= 2:
            self.write(PROTO + pack("<B", self.proto))
        if self.proto >= 4:
            self.framer.start_framing()
>       self.save(obj)

/data/paschalj/conda/envs/modindev10/lib/python3.6/pickle.py:409: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ray.cloudpickle.cloudpickle.CloudPickler object at 0x2aaab67175c0>, obj = <function PandasQueryCompiler._prepare_method.<locals>.helper at 0x2aaab671c7b8>, save_persistent_id = True

    def save(self, obj, save_persistent_id=True):
        self.framer.commit_frame()
    
        # Check for persistent id (defined by a subclass)
        pid = self.persistent_id(obj)
        if pid is not None and save_persistent_id:
            self.save_pers(pid)
            return
    
        # Check the memo
        x = self.memo.get(id(obj))
        if x is not None:
            self.write(self.get(x[0]))
            return
    
        # Check the type dispatch table
        t = type(obj)
        f = self.dispatch.get(t)
        if f is not None:
>           f(self, obj) # Call unbound method with explicit self

/data/paschalj/conda/envs/modindev10/lib/python3.6/pickle.py:476: 
```

jpdna commented:
I rebased before making further changes, but then got a segmentation fault I didn't get before.
To check that Master still works with my input, I went back and tried to use current Master on the input parquet file I had tried successfully earlier, and now get the following error using the Master branch:
```
In [2]: x5 = mpd.read_parquet("run3_num_only.100000.with_header.out.parquet")                                                                                                              
---------------------------------------------------------------------------
RayTaskError                              Traceback (most recent call last)
<ipython-input-2-d8ec8ed0a87c> in <module>
----> 1 x5 = mpd.read_parquet("run3_num_only.100000.with_header.out.parquet")

/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/modin-0.6.0.dev0-py3.6.egg/modin/pandas/io.py in read_parquet(path, engine, columns, **kwargs)
     27     return DataFrame(
     28         query_compiler=BaseFactory.read_parquet(
---> 29             path=path, columns=columns, engine=engine, **kwargs
     30         )
     31     )

/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/modin-0.6.0.dev0-py3.6.egg/modin/data_management/factories.py in read_parquet(cls, **kwargs)
     49     @classmethod
     50     def read_parquet(cls, **kwargs):
---> 51         return cls._determine_engine()._read_parquet(**kwargs)
     52 
     53     @classmethod

/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/modin-0.6.0.dev0-py3.6.egg/modin/data_management/factories.py in _read_parquet(cls, **kwargs)
     53     @classmethod
     54     def _read_parquet(cls, **kwargs):
---> 55         return cls.io_cls.read_parquet(**kwargs)
     56 
     57     @classmethod

/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/modin-0.6.0.dev0-py3.6.egg/modin/engines/ray/generic/io.py in read_parquet(cls, path, engine, columns, **kwargs)
    185             ]
    186         )
--> 187         index_len = ray.get(blk_partitions[-1][0])
    188         index = pandas.RangeIndex(index_len)
    189         new_query_compiler = cls.query_compiler_cls(

/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/ray-0.6.6-py3.6-linux-x86_64.egg/ray/worker.py in get(object_ids)
   2191                 # here.
   2192                 last_task_error_raise_time = time.time()
-> 2193                 raise value
   2194             return value
   2195 

RayTaskError: ray_worker (pid=51414, host=cn3811)
  File "/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/modin-0.6.0.dev0-py3.6.egg/modin/engines/ray/pandas_on_ray/io.py", line 54, in _read_parquet_columns
    df = pq.read_pandas(path, columns=columns, **kwargs).to_pandas()
  File "/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/ray-0.6.6-py3.6-linux-x86_64.egg/ray/pyarrow_files/pyarrow/parquet.py", line 1144, in read_pandas
    use_pandas_metadata=True)
  File "/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/ray-0.6.6-py3.6-linux-x86_64.egg/ray/pyarrow_files/pyarrow/parquet.py", line 1123, in read_table
    use_pandas_metadata=use_pandas_metadata)
  File "/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/ray-0.6.6-py3.6-linux-x86_64.egg/ray/pyarrow_files/pyarrow/filesystem.py", line 181, in read_parquet
    use_pandas_metadata=use_pandas_metadata)
  File "/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/ray-0.6.6-py3.6-linux-x86_64.egg/ray/pyarrow_files/pyarrow/parquet.py", line 985, in read
    use_pandas_metadata=use_pandas_metadata)
  File "/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/ray-0.6.6-py3.6-linux-x86_64.egg/ray/pyarrow_files/pyarrow/parquet.py", line 535, in read
    table = reader.read(**options)
  File "/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/ray-0.6.6-py3.6-linux-x86_64.egg/ray/pyarrow_files/pyarrow/parquet.py", line 211, in read
    columns, use_pandas_metadata=use_pandas_metadata)
  File "/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/ray-0.6.6-py3.6-linux-x86_64.egg/ray/pyarrow_files/pyarrow/parquet.py", line 261, in _get_column_indices
    for descr in index_columns
  File "/data/paschalj/conda/envs/jpmodin_5_13_2/lib/python3.6/site-packages/ray-0.6.6-py3.6-linux-x86_64.egg/ray/pyarrow_files/pyarrow/parquet.py", line 262, in <listcomp>
    if descr['kind'] == 'serialized']
TypeError: string indices must be integers
```

so some commit to Master broke reading of my input parquet file over the last month.
I'll look into fiurther, but let me know if you have ideas.

jpdna commented:
I'll right, well using current Master I can write and then re-read my test parquet file starting from a txt input, so I'll work from there.

jpdna commented:
closed in favor of #664
